### PR TITLE
JAMES-3179 Result of a static analysis results

### PR DESCRIPTION
**Pull request by blackheaven**

### CODE REVIEW

File: mailet/mailetdocs-maven-plugin/src/main/java/org/apache/james/mailet/DefaultDescriptorsExtractor.java [hunk 1]
```diff
@@ -163,27 +164,34 @@ private MailetMatcherDescriptor describeMatcher(Log log,
...
-    private MailetMatcherDescriptor buildDescriptor(Log log, JavaClass nextClass,
-            final String nameOfClass, Class<?> klass,
-            final String infoMethodName, Type type) {
-        final MailetMatcherDescriptor result = new MailetMatcherDescriptor();
-        result.setName(nextClass.getName());
-        result.setFullyQualifiedName(nameOfClass);
-        result.setClassDocs(nextClass.getComment());
-        result.setType(type);
-        result.setExperimental(isExperimental(nextClass));
+    private MailetMatcherDescriptor buildDescriptor(Log log,
```
Comment (rouazana) [inline mailet/mailetdocs-maven-plugin/src/main/java/org/apache/james/mailet/DefaultDescriptorsExtractor.java:167]: is this method still needed?

File: protocols/imap/src/main/java/org/apache/james/imap/message/request/AbstractMessageRangeRequest.java [hunk 2]
```diff
@@ -74,6 +74,6 @@ private boolean equals(IdRange[] idSet1, IdRange[] idSet2) {
 
     @Override
     public int hashCode() {
-        return Objects.hashCode(idSet, mailboxName, useUids);
+        return Objects.hashCode(Arrays.hashCode(idSet), mailboxName, useUids);
```
Comment (rouazana) [inline protocols/imap/src/main/java/org/apache/james/imap/message/request/AbstractMessageRangeRequest.java:77]: no equalsverifier?

File: mpt/core/src/test/java/org/apache/james/mpt/api/ImapFeaturesTest.java [hunk 3]
```diff
@@ -62,13 +62,15 @@ public void supportsShouldReturnTrueOnEmptyListWhenNoFeatures() {
     }
 
     @Test(expected = NullPointerException.class)
+    @SuppressWarnings("CheckReturnValue")
     public void supportsShouldThrowOnNullFeature() {
-        assertThat(ImapFeatures.of().supports((Feature)null));
+        ImapFeatures.of().supports((Feature)null);
```
Comment (rouazana) [inline mpt/core/src/test/java/org/apache/james/mpt/api/ImapFeaturesTest.java:67]: assertThatThrows should remove this warning

File: mpt/core/src/test/java/org/apache/james/mpt/api/ImapFeaturesTest.java [hunk 4]
```diff
@@ -62,13 +62,15 @@ public void supportsShouldReturnTrueOnEmptyListWhenNoFeatures() {
...
     }
 
     @Test(expected = NullPointerException.class)
+    @SuppressWarnings("CheckReturnValue")
     public void supportsShouldThrowOnNullFeatureArray() {
-        assertThat(ImapFeatures.of().supports((Feature[])null));
+        ImapFeatures.of().supports((Feature[])null);
```
Comment (rouazana) [inline mpt/core/src/test/java/org/apache/james/mpt/api/ImapFeaturesTest.java:73]: idem

File: server/data/data-library/src/test/java/org/apache/james/rrt/lib/MappingsImplTest.java [hunk 5]
```diff
@@ -158,9 +158,10 @@ public void containsShouldReturnFalseWhenNoMatchingMapping() {
 
     
     @Test(expected = NullPointerException.class)
+    @SuppressWarnings("CheckReturnValue")
     public void containsShouldThrowWhenNull() {
         MappingsImpl mappings = MappingsImpl.builder().add(Mapping.regex("toto")).build();
-        assertThat(mappings.contains((Mapping.Type)null));
+        mappings.contains((Mapping.Type)null);
```
Comment (rouazana) [inline server/data/data-library/src/test/java/org/apache/james/rrt/lib/MappingsImplTest.java:164]: idem

File: server/data/data-library/src/test/java/org/apache/james/rrt/lib/MappingsImplTest.java [hunk 6]
```diff
@@ -190,9 +191,10 @@ public void selectShouldReturnEmptyCollectionWhenNoMatchingMapping() {
 
     
     @Test(expected = NullPointerException.class)
+    @SuppressWarnings("CheckReturnValue")
     public void selectShouldThrowWhenNull() {
         MappingsImpl mappings = MappingsImpl.builder().add(Mapping.regex("toto")).build();
-        assertThat(mappings.select((Mapping.Type)null));
+        mappings.select((Mapping.Type)null);
```
Comment (rouazana) [inline server/data/data-library/src/test/java/org/apache/james/rrt/lib/MappingsImplTest.java:197]: idem

File: server/data/data-library/src/test/java/org/apache/james/rrt/lib/MappingsImplTest.java [hunk 7]
```diff
@@ -221,9 +223,10 @@ public void excludeShouldReturnEmptyCollectionWhenOnlyMatchingMapping() {
 
     
     @Test(expected = NullPointerException.class)
+    @SuppressWarnings("CheckReturnValue")
     public void excludeShouldThrowWhenNull() {
         MappingsImpl mappings = MappingsImpl.builder().add(Mapping.regex("toto")).build();
-        assertThat(mappings.exclude((Mapping.Type)null));
+        mappings.exclude((Mapping.Type)null);
```
Comment (rouazana) [inline server/data/data-library/src/test/java/org/apache/james/rrt/lib/MappingsImplTest.java:229]: idem

File: pom.xml [hunk 8]
```diff
@@ -2726,6 +2726,26 @@
     <build>
         <pluginManagement>
             <plugins>
+                <plugin>
```
Comment (chibenwa) [inline pom.xml:2729]: Can we have a documentation somewhere of how to use this plugin?

File: backends-common/jpa/src/main/java/org/apache/james/backends/jpa/TransactionRunner.java [hunk 9]
```diff
@@ -76,8 +76,8 @@ public void run(Consumer<EntityManager> runnable) {
         }
```
Comment (chibenwa) [inline backends-common/jpa/src/main/java/org/apache/james/backends/jpa/TransactionRunner.java:76]: > JAMES-3179 Use a Consumer instead of a Function in TransactionRunner 

Why?

(Whithout an explanation I might do the mistake again)

File: mailbox/maildir/src/test/java/org/apache/james/mailbox/maildir/OsDetector.java [hunk 10]
```diff
@@ -29,8 +30,10 @@
      * @return windows
      */
     public static boolean isWindows() {
-        String os = System.getProperty("os.name").toLowerCase(Locale.US);
-        return (os.contains("win"));
+        return Optional.ofNullable(System.getProperty("os.name"))
+            .orElse("")
+            .toLowerCase(Locale.US)
+            .contains("win");
```
Comment (chibenwa) [inline mailbox/maildir/src/test/java/org/apache/james/mailbox/maildir/OsDetector.java:36]: No tests?

(we could extract a submethod taking a `Supplier<String> osSupplier` and test it)

File: protocols/imap/src/main/java/org/apache/james/imap/decode/ImapRequestStreamLineReader.java [hunk 11]
```diff
@@ -102,4 +102,9 @@ protected void commandContinuationRequest() throws DecodingException {
         }
     }
 
+    @Override
+    public void close() throws IOException {
+        input.close();
+        output.close();
```
Comment (chibenwa) [inline protocols/imap/src/main/java/org/apache/james/imap/decode/ImapRequestStreamLineReader.java:108]: OUCH! Well spotted!

**Comment by nvduc91** [comment]
read it.

File: pom.xml [hunk 12]
```diff
@@ -2726,6 +2726,26 @@
...
+                    <groupId>org.apache.maven.plugins</groupId>
+                    <artifactId>maven-compiler-plugin</artifactId>
+                    <version>3.8.0</version>
+                    <configuration>
+                        <source>8</source>
```
Comment (mbaechler) [inline pom.xml:2734]: should be 11 right?

File: pom.xml [hunk 13]
```diff
@@ -2726,6 +2726,26 @@
...
+                        <target>8</target>
+                        <compilerArgs>
+                            <arg>-XDcompilePolicy=simple</arg>
+                            <arg>-Xplugin:ErrorProne</arg>
```
Comment (mbaechler) [inline pom.xml:2738]: what is the impact on the compile time?

File: backends-common/jpa/src/main/java/org/apache/james/backends/jpa/TransactionRunner.java (hunk 9 above)
Comment (mbaechler) [reply backends-common/jpa/src/main/java/org/apache/james/backends/jpa/TransactionRunner.java:76]: we don't use the return value?

File: mailet/mailetdocs-maven-plugin/src/main/java/org/apache/james/mailet/AbstractMailetdocsReport.java [hunk 14]
```diff
@@ -199,17 +199,14 @@ private void outputDescriptorList(List<MailetMatcherDescriptor> descriptors, Str
                     getSink().text("Info: ");
                 }
                 getSink().bold();
-                getSink().text(descriptor.getInfo());
+                getSink().text(descriptor.getInfo().orElse(""));
```
Comment (mbaechler) [inline mailet/mailetdocs-maven-plugin/src/main/java/org/apache/james/mailet/AbstractMailetdocsReport.java:202]: does the null check some lines above still make sense?

File: protocols/imap/src/test/java/org/apache/james/imap/decode/DecoderUtilsLocaleDateTest.java [hunk 15]
```diff
@@ -175,7 +175,7 @@ void decodeDateTimeWithVariousDefaultLocale(String input, String expected, Local
         LocalDateTime localDateTime = DecoderUtils.decodeDateTime(input);
         assertThat(ZonedDateTime.of(localDateTime, ZoneId.systemDefault())
             .withZoneSameInstant(ZoneId.of("GMT"))
-            .format(DateTimeFormatter.ofPattern("dd MMM YYYY HH:mm:ss VV", Locale.US)))
+            .format(DateTimeFormatter.ofPattern("dd MMM yyyy HH:mm:ss VV", Locale.US)))
```
Comment (mbaechler) [inline protocols/imap/src/test/java/org/apache/james/imap/decode/DecoderUtilsLocaleDateTest.java:178]: no test?

File: server/container/spring/src/main/java/org/apache/james/container/spring/osgi/AbstractBundleTracker.java [hunk 16]
```diff
@@ -88,7 +88,7 @@ public void bundleChanged(BundleEvent event) {
                     URL e = (URL) entrs.nextElement();
                     String file = e.getFile();
 
-                    String className = file.replaceAll("/", ".").replaceAll(".class", "").replaceFirst(".", "");
+                    String className = file.replaceAll("/", ".").replaceAll("\\.class", "").replaceFirst("\\.", "");
```
Comment (mbaechler) [inline server/container/spring/src/main/java/org/apache/james/container/spring/osgi/AbstractBundleTracker.java:91]: no test?

File: protocols/imap/src/main/java/org/apache/james/imap/processor/StoreProcessor.java [hunk 17]
```diff
@@ -120,9 +120,6 @@ protected void processRequest(StoreRequest request, ImapSession session, Respond
                 if (messageSet != null) {
 
                     if (unchangedSince != -1) {
-                        // Ok we have a CONDSTORE option so use the CONDSTORE_COMMAND
-                        imapCommand = CONDSTORE_COMMAND;
```
Comment (mbaechler) [inline protocols/imap/src/main/java/org/apache/james/imap/processor/StoreProcessor.java:124]: it looks weird, why you need to remove that?

File: protocols/imap/src/main/java/org/apache/james/imap/decode/ImapRequestStreamLineReader.java [hunk 18]
```diff
@@ -102,4 +102,9 @@ protected void commandContinuationRequest() throws DecodingException {
         }
     }
 
+    @Override
+    public void close() throws IOException {
+        input.close();
```
Comment (mbaechler) [inline protocols/imap/src/main/java/org/apache/james/imap/decode/ImapRequestStreamLineReader.java:107]: what if this line throws?

File: server/data/data-file/src/main/java/org/apache/james/sieverepository/file/SieveFileRepository.java [hunk 19]
```diff
@@ -199,25 +203,25 @@ public void haveSpace(Username username, ScriptName name, long size) throws Quot
...
         }
-        for (File file : files) {
-            if (!SYSTEM_FILES.contains(file.getName())) {
-                summaries.add(new ScriptSummary(new ScriptName(file.getName()), isActive(file, activeFile)));
-            }
-        }
-        return summaries;
+
+        Predicate<File> isActive = isActiveValidator(activeFile);
+        return Stream.of(Optional.ofNullable(getUserDirectory(username).listFiles()).orElse(new File[]{}))
```
Comment (mbaechler) [inline server/data/data-file/src/main/java/org/apache/james/sieverepository/file/SieveFileRepository.java:214]: no test?

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/ResultUtils.java [hunk 20]
```diff
@@ -216,9 +216,11 @@ private static void addBodyContent(MailboxMessage message, MessageResultImpl mes
     private static void addMimeBodyContent(MailboxMessage message, MessageResultImpl messageResult, MimePath mimePath)
             throws IOException, MimeException {
         int[] path = path(mimePath);
-        PartContentBuilder builder = build(path, message);
-        Content content = builder.getMimeBodyContent();
-        messageResult.setMimeBodyContent(mimePath, content);
+        if (path != null) {
```
Comment (mbaechler) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/ResultUtils.java:219]: no test?

File: protocols/imap/src/main/java/org/apache/james/imap/message/request/AbstractMessageRangeRequest.java (hunk 2 above)
Comment (blackheaven) [reply protocols/imap/src/main/java/org/apache/james/imap/message/request/AbstractMessageRangeRequest.java:77]: I do not know how to do it on an abstract class

File: pom.xml (hunk 8 above)
Comment (blackheaven) [reply pom.xml:2729]: It is quite straightforward: if it finds an error, the compilation breaks

File: backends-common/jpa/src/main/java/org/apache/james/backends/jpa/TransactionRunner.java (hunk 9 above)
Comment (blackheaven) [reply backends-common/jpa/src/main/java/org/apache/james/backends/jpa/TransactionRunner.java:76]: In fact, we require a `Function` but we ignore the result, making the interface suspicious

File: protocols/imap/src/test/java/org/apache/james/imap/decode/DecoderUtilsLocaleDateTest.java (hunk 15 above)
Comment (blackheaven) [reply protocols/imap/src/test/java/org/apache/james/imap/decode/DecoderUtilsLocaleDateTest.java:178]: no test test? :)

File: protocols/imap/src/main/java/org/apache/james/imap/processor/StoreProcessor.java (hunk 17 above)
Comment (blackheaven) [reply protocols/imap/src/main/java/org/apache/james/imap/processor/StoreProcessor.java:124]: it was only set, never read

File: protocols/imap/src/main/java/org/apache/james/imap/decode/ImapRequestStreamLineReader.java (hunk 18 above)
Comment (blackheaven) [reply protocols/imap/src/main/java/org/apache/james/imap/decode/ImapRequestStreamLineReader.java:107]: the caller deals with it (and ignore it currently)

File: server/container/spring/src/main/java/org/apache/james/container/spring/osgi/AbstractBundleTracker.java (hunk 16 above)
Comment (blackheaven) [reply server/container/spring/src/main/java/org/apache/james/container/spring/osgi/AbstractBundleTracker.java:91]: It's hard to test an abstract class

File: pom.xml (hunk 8 above)
Comment (chibenwa) [reply pom.xml:2729]: Ok thanks, it is runned by default

File: server/data/data-file/src/main/java/org/apache/james/sieverepository/file/SieveFileRepository.java (hunk 19 above)
Comment (blackheaven) [reply server/data/data-file/src/main/java/org/apache/james/sieverepository/file/SieveFileRepository.java:214]: I should create a fs with wrong credentials, it looks overrated

File: pom.xml (hunk 13 above)
Comment (blackheaven) [reply pom.xml:2738]:
```
Without:
mvn package -T0.5C -DskipTests  2116.88s user 120.28s system 681% cpu 5:28.31 total
With:
mvn package -T0.5C -DskipTests  3050.80s user 131.11s system 685% cpu 7:44.38 total
```

Too much, I'll drop it

File: mailbox/maildir/src/test/java/org/apache/james/mailbox/maildir/OsDetector.java (hunk 10 above)
Comment (blackheaven) [reply mailbox/maildir/src/test/java/org/apache/james/mailbox/maildir/OsDetector.java:36]: it is not used

File: server/container/spring/src/main/java/org/apache/james/container/spring/osgi/AbstractBundleTracker.java (hunk 16 above)
Comment (mbaechler) [reply server/container/spring/src/main/java/org/apache/james/container/spring/osgi/AbstractBundleTracker.java:91]: btw I see it"s located in an osgi package, is it used at all?

File: protocols/imap/src/main/java/org/apache/james/imap/decode/ImapRequestStreamLineReader.java (hunk 18 above)
Comment (mbaechler) [reply protocols/imap/src/main/java/org/apache/james/imap/decode/ImapRequestStreamLineReader.java:107]: but the next line never execute

File: mpt/core/src/test/java/org/apache/james/mpt/api/ImapFeaturesTest.java [hunk 21]
```diff
@@ -19,68 +19,71 @@
...
+            .isInstanceOf(NullPointerException.class);
     }
 
-    @Test(expected = NullPointerException.class)
-    @SuppressWarnings("CheckReturnValue")
-    public void supportsShouldThrowOnNullFeatureArray() {
-        ImapFeatures.of().supports((Feature[])null);
+    @Test
+    void supportsShouldThrowOnNullFeatureArray() {
+        assertThatCode(() -> ImapFeatures.of().supports((Feature[])null))
```
Comment (mbaechler) [inline mpt/core/src/test/java/org/apache/james/mpt/api/ImapFeaturesTest.java:73]: it should be `assertThatThrownBy`

File: pom.xml (hunk 13 above)
Comment (mbaechler) [reply pom.xml:2738]: could be out of `compile` stage, we could group that stage with checkstyle

File: server/container/spring/src/main/java/org/apache/james/container/spring/osgi/AbstractBundleTracker.java (hunk 16 above)
Comment (blackheaven) [reply server/container/spring/src/main/java/org/apache/james/container/spring/osgi/AbstractBundleTracker.java:91]: I have no idea, should I drop all the file?

File: pom.xml (hunk 13 above)
Comment (blackheaven) [reply pom.xml]: I have tried something, I am not sure it is what you want, pairing on it would be great

File: server/container/spring/src/main/java/org/apache/james/container/spring/osgi/AbstractBundleTracker.java (hunk 16 above)
Comment (mbaechler) [reply server/container/spring/src/main/java/org/apache/james/container/spring/osgi/AbstractBundleTracker.java:91]: remove it and see if it still pass the CI?

File: pom.xml (hunk 13 above)
Comment (mbaechler) [reply pom.xml]: Looking at https://maven.apache.org/guides/introduction/introduction-to-the-lifecycle.html#lifecycle-reference there's not phase that would fit between compile and verify.

Either we enable it at test stage or at compile stage or we wait the switch to gradle (no ETA at all on that)

**Comment by Arsnael** [comment]
Lost output logs

test this please

**Comment by chibenwa** [comment]
test this please (build result lost)

File: pom.xml (hunk 13 above)
Comment (blackheaven) [reply pom.xml]: let's drop it then

**Comment by blackheaven** [comment]
no results

test this please

**Comment by blackheaven** [comment]
```
[b1ead95cdca8a3485031714a82107a2e199ee950] Exception in thread "Thread-675" java.lang.RuntimeException: java.io.IOException: Failed to bind to /0.0.0.0:43197
[b1ead95cdca8a3485031714a82107a2e199ee950] 	at org.apache.james.webadmin.WebAdminServer.lambda$start$1(WebAdminServer.java:95)
```

test this please

**Comment by blackheaven** [comment]
I think I have fixed IMAP:

```
Expected: '1e NO \[TRYCREATE\] APPEND failed. No such mailbox.'
Actual   : '* 2 EXISTS'
```

**Comment by chibenwa** [comment]
> I think I have fixed IMAP:

Can you develop a bit more? Open a related ticket maybe?

**Comment by blackheaven** [comment]
Wrong ticket: https://github.com/linagora/james-project/pull/3281#issuecomment-634474425

File: server/container/spring/src/main/java/org/apache/james/container/spring/osgi/AbstractBundleTracker.java (hunk 16 above)
Comment (blackheaven) [reply server/container/spring/src/main/java/org/apache/james/container/spring/osgi/AbstractBundleTracker.java]: not used => dropping

File: server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/crypto/SecurityKeyLoader.java [hunk 22]
```diff
@@ -55,9 +55,11 @@ public AsymmetricKeys load() throws Exception {
         Preconditions.checkState(jmapDraftConfiguration.isEnabled(), "JMAP is not enabled");
 
         KeyStore keystore = KeyStore.getInstance(JKS);
-        InputStream fis = fileSystem.getResource(jmapDraftConfiguration.getKeystore());
-        char[] secret = jmapDraftConfiguration.getSecret().toCharArray();
-        keystore.load(fis, secret);
+        char[] secret;
+        try (InputStream fis = fileSystem.getResource(jmapDraftConfiguration.getKeystore())) {
+            secret = jmapDraftConfiguration.getSecret().toCharArray();
```
Comment (chibenwa) [inline server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/crypto/SecurityKeyLoader.java:60]: This should be handled in another PR (clearly) but we don't zero the secret char array after using it?

Comment (chibenwa) [reply server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/crypto/SecurityKeyLoader.java:60]: Answer, the password is in a String just above, and we are forced by configuration2 to access the secret as a String...

**Comment by rouazana** [comment]
merged
